### ISSUE TITLE
[Proposal] New `info` API for vectorized environments

### ISSUE THREAD

**Issue by vwxyzjn** [openai/gym#2657]
### Motivation

The current vector env API would return an `info` dict per environment. So if  if I have 5 envs, then the dict could look like
```[{}, {}, {}, {}, {"truncated": True}]```

While this API is intuitive, it's not a friendly interface to high-throughput environments such as [brax](https://github.com/google/brax) (also see https://github.com/google/brax/issues/164) and [envpool](https://github.com/sail-sg/envpool). In these envs, they prefer the following paradigm: instead of 5 dictionaries, it's just a single dictionary with an array of length 5.

```{"truncated":[False, False, False, False, True]}```

## Pros and Cons

Brax and envpool's approach feels more ergonomic for high-throughput envs, whereas Gym's approach is maybe more efficient for "sparse" info keys that don't appear often (I could be wrong with this). 

### Proposal 

I propose to change the `info` API for vectorized environments to adapt brax's current design for `info`.  This might also have implications for wrappers and other APIs. For example, the current vectorized environment API would return `info` like this (with the `RecordEpisodeStatistics` used):

```python
[
    {
        "episode": {
            "r": 75.0,
            "l": 75,
            "t": 36.770559,
        },
        "terminal_observation": array([0.2097086, -0.5355723, -0.21343598, -0.11173592], dtype=float32),
    },
    {},
    {},
    {},
    {},
]
```

I am not sure what's the best way to handle this... Maybe something like the following?

```python
{
    "episode": {
        "r": [75.0, None, None, None, None],
        "l": [75, None, None, None, None],
        "t": [36.770559, None, None, None, None],
    },
    "terminal_observation": [
        array([0.2097086, -0.5355723, -0.21343598, -0.11173592], dtype=float32),
        None,
        None,
        None,
        None,
    ]
}
```

### Alternative 1

Alternatively, it's possible to do this in a very hacky way: brax or env pool could always override the first sub-env's info dict to always provide their desired info so something like:

```[{"truncated":[False, False, False, False, True]}, {}, {}, {}, {}]```

This way, we don't have to change the current API, however, it's extremely hacky and unintuitive.

### Alternative 2

Alternatively, maybe `info` should be reserved for something sparse like `episode` and `terminal_observation`. And for more frequent infos they should be accessed via attributes (i.e., `envs.get_attr(name)` (enabled with https://github.com/openai/gym/pull/1600). So something like

```python
root_joints = brax_envs.get_attr("root_joints")
# instead of 
# root_joints = info["root_joints"]
ale_lives = envpool_envs.get_attr("ale_lives")
# instead of 
# ale_lives = info["ale_lives"]
```

What are your thoughts @RedTachyon @JesseFarebro, @tristandeleu @XuehaiPan @Trinkle23897 @erikfrey @araffin @Miffyli

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/openai/gym/issues) in the repo (**required**)

**Comment by Trinkle23897** [comment]
I'd suggest not to use None because it's hard to directly convert the array into tensor. In general, tianshou's Batch has already proposed a potential solution like:
```python
In [24]: from tianshou.data import Batch

In [25]: a = Batch(episode={'l': 75, 'r': 75.0, 't': 36.77})

In [26]: b = Batch(terminal_obs=0.2097)

In [27]: Batch.stack([a, b])
Out[27]: 
Batch(
    terminal_obs: array([0.    , 0.2097]),
    episode: Batch(
                 l: array([75,  0]),
                 r: array([75.,  0.]),
                 t: array([36.77,  0.  ]),
             ),
)
```

**Comment by gianlucadecola** [comment]
I'm currently looking at this; for wrappers like ```RecordEpisodeStatistics``` the most reasonable approach to me seems the first proposal
```
{
    "episode": {
        "r": [75.0, None, None, None, None],
        "l": [75, None, None, None, None],
        "t": [36.770559, None, None, None, None],
    },
    "terminal_observation": [
        array([0.2097086, -0.5355723, -0.21343598, -0.11173592], dtype=float32),
        None,
        None,
        None,
        None,
    ]
}
```
Since there are a lot of things to work on I think it is better to agree on the new API before starting to make changes.

Also the solution proposed by @Trinkle23897 seems good, my only doubt is that using 0's can be misleading regarding the fact that it is a "real" zero or a zero from the fact that there is no info for the env.

**Comment by pseudo-rnd-thoughts** [comment]
There is a third option which I suspect to be inefficient however removes the need for no-data data, i.e. `0` or `None`.
That is pairing the info data with the environment number.
This I think would be efficient for sparse environment info but not for dense
```
{
   "key": {
         (data, environment number) 
         ...
   }, 
   ...
}
```

Otherwise I think that brax and tianshou have very similar implementations that we could follow making gym work with their modules easier.

**Comment by jkterry1** [comment]
Closing in favor of PR
